lib/ABI/FunctionType.cpp

```diff
@@ -727,31 +725,46 @@ class ConversionHelper {
   distributeNonPositionBasedArguments(const ArgumentContainer &Arguments,
                                       std::size_t SkippedRegisters = 0) {
     DistributedArguments Result;
-    size_t UsedGeneralPurposeRegisterCounter = SkippedRegisters;
-    size_t UsedVectorRegisterCounter = 0;
+    size_t UsedGeneralPurposeRegisterCount = SkippedRegisters;
+    size_t UsedVectorRegisterCount = 0;
 
     for (const model::Argument &Argument : Arguments) {
       auto MaybeSize = Argument.Type.size();
       revng_assert(MaybeSize.has_value());
 
       constexpr bool CanSplit = AT::ArgumentsCanBeSplitBetweenRegistersAndStack;
       if (Argument.Type.isFloat()) {
+        // The conventional non-position based approach is not applicable for
+        // vector registers since it's rare for multiple registers to be used
+        // to pass a single argument.
+        //
+        // For now, just return the next free register.
+        //
+        // TODO: handle this properly.
```

Comment:
I don't understand. Can you make an example of what situation we're not handling?

Reply:
For example, `float128`s under `s390x` are supposed to be returned on the stack instead of the register.

Same for more typical architectures. If, for example under SystemV_x64, you try to return `float128` on an architecture that doesn't have vector registers big enough to hold it (most of the floating point stuff are declared as extensions after all), the behaviour should be the same, as in - use stack.

But since for now we have no way to declare something as specific as available extensions (within the model, abi trait or whatnot), I decided to just go with "handle the most common situation, fail miserable with the rest" plus "leave a TODO" approach.

Reply:
Uhm, I think the best approach would be to assume a certain extension being present and that's it.
In the future, we'll enable variants, but for now I think we should just pick one and stick with it.

For instance, as of now, we assume we have `xmm0`, so we're already choosing an ABI, I think?

Reply:
But the "real" size of the `xmm` is also extension dependent. It could be 128 bit (`xmm`), or 256 (`ymm`), or 512 (`zmm`). If we try to return a value we have no register for under the current architecture (including extensions), we return it on the stack instead.

Reply:
Also, under the `SystemV_x86_64` there is a `long double` (80-bit) beast which is returned in `stN` registers :cry:

Reply:
We say `xmm` so we assume 128-bits in `Register.h`. Let's stick with that.

> Also, under the SystemV_x86_64 there is a long double (80-bit) beast which is returned in stN (x87) registers cry

I would say we emit a message on the logger and pretend that return value doesn't exist.
If it's not in `Register.h`, we're ignoring it anyway.

Reply:
Since this is still unresolved when the branch got merged, let's move it to #232. I'll recreate the thread there and link to this one, allowing this one to get closed.